calico/felix/test/test_fetcd_api.py

```diff
@@ -0,0 +1,185 @@
+    @patch('calico.felix.fetcd._EtcdWatcher')
+    @patch('calico.felix.fetcd.etcd')
+    def finish_setup(self, m_etcd, m_EtcdWatcher, m_write_to_etcd, **kwargs):
+        # Set configuration attributes and start etcd_api
+        for key, value in kwargs.iteritems():
+            setattr(self.m_config, key, value)
+
+        self.m_hosts_ipset = Mock()
+        self.etcd_api = EtcdAPI(self.m_config, self.m_hosts_ipset)
+
+        self.etcd_api.write_to_etcd = Mock()
+
+    def run_actor_loop(self):
+        self.etcd_api._step()
+
+    def test_write_to_etcd_actor_message(self):
+        """
+        Test write_to_etcd actor message calls client.write
+        """
+        self.m_hosts_ipset = Mock()
+        self.etcd_api = EtcdAPI(self.m_config, self.m_hosts_ipset)
+
+        with patch.object(self.etcd_api.client, 'write'):
+            self.etcd_api.write_to_etcd('key', 'value', async=True)
+            self.run_actor_loop()
+            self.assertTrue(self.etcd_api.client.write.called)
+
+    def test_update_felix_status_disabled(self):
+        """
+        Test reporting is disabled for reporting interval 0
+        """
+        self.finish_setup()
+        # We call gevent.sleep so that _status_reporting thread is executed -
+        # since no interval and ttl is given, _status_reporting thread returns
+        # instead of writing to etcd.
+        gevent.sleep(1)
+        self.assertFalse(self.etcd_api.write_to_etcd.called)
+
+    def test_update_felix_status_single(self):
+        """
+        Test felix status is updated
+        """
+        self.finish_setup(REPORTING_INTERVAL_SECS=15,
+                          REPORTING_TTL_SECS=37)
+
+        hostname = self.etcd_api._config.HOSTNAME
+        status_key = key_for_status(hostname)
+        uptime_key = key_for_uptime(hostname)
+        ttl = self.etcd_api._config.REPORTING_TTL_SECS
+
+        # We call gevent.sleep so that _status_reporting thread is executed
+        gevent.sleep(1)
+
+        status_call = call(status_key, TestIfStatus(), async=True)
+        uptime_call = call(uptime_key, TestIfUptime(), ttl=ttl, async=True)
+        self.etcd_api.write_to_etcd.assert_has_calls([status_call,
+                                                      uptime_call])
+
+    def test_update_felix_status_continuous(self):
+        """
+        Test felix status is being continuously updated
+        """
+        self.finish_setup(REPORTING_INTERVAL_SECS=5,
+                          REPORTING_TTL_SECS=17)
+        hostname = self.etcd_api._config.HOSTNAME
+        status_key = key_for_status(hostname)
+        uptime_key = key_for_uptime(hostname)
+        ttl = self.etcd_api._config.REPORTING_TTL_SECS
+
+        status_call = call(status_key, TestIfStatus(), async=True)
+        uptime_call = call(uptime_key, TestIfUptime(), ttl=ttl, async=True)
+
+        # We call gevent.sleep so that _status_reporting thread is executed
+        gevent.sleep(5)

```

Comment:
This test will take at least 10s if I'm reading it right.  That's *far* too long.  Right now all the tests run in 12s and that really reduces development cycles.  You need to find a way to avoid needing to sleep.